### ISSUE TITLE
Intermittent "Failed to build Object (error 0x1)" errors, triggered by remote job beating local job in race

### ISSUE THREAD

**Issue by nedwardsnae** [fastbuild/fastbuild#486]
Hi,

We've been getting intermittent build failures for some time now with our UE4 project. Having finally gotten around to looking at this in more detail, I noticed we had a number of FBuild.exe CrashDumps. They seem to coincide with our build failures, so I suspect that the reason for the crash is the reason for why our builds fail.

Having done some investigation, all of the dumps report crashing due to stack overflow:
```
 	kernel32.dll!Process32First()	Unknown
 	FBuild.exe!Process::KillProcessTreeInternal(unsigned int processID) Line 87	C++
 	...
 	FBuild.exe!Process::KillProcessTreeInternal(unsigned int processID) Line 99	C++
 	FBuild.exe!Process::KillProcessTreeInternal(unsigned int processID) Line 99	C++
 	FBuild.exe!Process::KillProcessTreeInternal(unsigned int processID) Line 99	C++
 	FBuild.exe!Process::KillProcessTreeInternal(unsigned int processID) Line 99	C++
 	FBuild.exe!Process::KillProcessTreeInternal(unsigned int processID) Line 99	C++
 	FBuild.exe!Process::KillProcessTreeInternal(unsigned int processID) Line 99	C++
 	[Inline Frame] FBuild.exe!Process::KillProcessTree() Line 131	C++
 	FBuild.exe!Process::ReadAllData(AutoPtr<char,DefaultDeletor> & outMem, unsigned int * outMemSize, AutoPtr<char,DefaultDeletor> & errMem, unsigned int * errMemSize, unsigned int timeOutMS) Line 542	C++
 	FBuild.exe!ObjectNode::CompileHelper::SpawnCompiler(Job * job, const AString & name, const AString & compiler, const Args & fullArgs, const char * workingDir) Line 2192	C++
 	FBuild.exe!ObjectNode::BuildFinalOutput(Job * job, const Args & fullArgs) Line 2112	C++
 	FBuild.exe!ObjectNode::DoBuildWithPreProcessor2(Job * job, bool useDeoptimization, bool stealingRemoteJob, bool racingRemoteJob) Line 538	C++
 	FBuild.exe!ObjectNode::DoBuild2(Job * job, bool racingRemoteJob) Line 285	C++
 	FBuild.exe!JobQueueRemote::DoBuild(Job * job, bool racingRemoteJob) Line 324	C++
>	[Inline Frame] FBuild.exe!WorkerThread::Update() Line 205	C++
 	FBuild.exe!WorkerThread::Main() Line 132	C++
 	FBuild.exe!WorkerThread::ThreadWrapperFunc(void * param) Line 113	C++
 	kernel32.dll!BaseThreadInitThunk()	Unknown
 	ntdll.dll!RtlUserThreadStart()	Unknown
```

The issue seems to be that a remote job finishes first, causing the local job to be aborted. When the local job is aborted, it attempts to kill the process, but gets stuck trying to find child processes of the main process. This is because cl.exe and FBuild.exe are each other's parent...yep:
![devenv_2018-09-04_15-51-10](https://user-images.githubusercontent.com/3097360/45044499-33fb6180-b068-11e8-99c5-d9dd6d9092bd.png)
![devenv_2018-09-04_15-51-25](https://user-images.githubusercontent.com/3097360/45044500-33fb6180-b068-11e8-9f67-064cc61cdee9.png)
I'm not sure how this happens, but it may explain why we get these intermittent failures.

Other than disabling local races, does anyone know what could be going on here? I'm pretty puzzled by it.

We're using FastBuild v0.96.

Thanks,
Nick

**Comment by minid33** [comment]
To add some context, this issue occurs to us on 2 separate build clusters which have no interaction with each other.

**Comment by ffulin** [comment]
That's very strange.  It would go some way to explaining the mysterious error 0x1 though: If when attempting to kill a child cl.exe, we think FASTBuild is a child of that, then we would in turn think that all other cl.exes for other threads are also children and we probably kill those.

I don't understand how this cycle in the process tree can exist though....

**Comment by ffulin** [comment]
I'd love to understand the root of this issue, but a potential work around could be to modify the process walking code to ignore sub trees in the graph that contains fbuild.exe's processId. That should avoid the stack overflow as well as potentially avoiding terminating the wrong child process.

**Comment by jspelletier** [comment]
Hi Franta, 

We also got this bug and we're pretty sure  that the problem is caused by recycling of PID for the intermediate fbuild process. Due to Windows recycling of pids, sometimes, a cl.exe gets the same PID than the one that was assigned to the intermediate fbuild.exe. This is causing inconsistencies in the process tree.

We've done two fixes for this bug internally at Ubisoft this week. We will run with those fixes for a while and then push some merge requests if everything is fine.
- The first one avoid PID recycling for the intermediate process by keeping keeping handle to the process around which blocks Windows from recycling that PID even if the process has exited.
- The second one changes the kill process logic to avoid any possible stack overflow. It also excludes killing the fbuild process(failsafe).

Cheers.
JS

**Comment by ffulin** [comment]
I've been thinking about this problem, and brainstorming with a colleague, we came to the same conclusion that PID reuse, combined with the "wrapper" process no longer existing could mean that a child process (like a cl.exe) can have the same PID as the process that at one point was the parent of the primary fbuild.exe. This causes the looping/terminating the wrong thing/stack overflow crazyness we're seeing.

Thinking about how to robustly fix this, I think what we need to do is query the start time of every process we see as we descend the tree, and only terminate children and continue descending when the creation time >= that of the parent process. I think GetProcessTimes() will give us what we need.

I think that handles the PID reuse without the need for an extra lock, as well as providing a failsafe upper bound. It also handles any case where a 3rd party program might start fbuild.exe and exit, leaving us with the same PID re-use problem but across unrelated process hierarchies.

**Comment by bburgin** [comment]
I am working on a fix for this issue that mimics what Jenkins does for killing child processes:
Add an env variable to children as you spawn them, and only kill processes with the specific env variable value. This has the added benefit that grandchild processes inherit the variable value and so the parent process can kill all child processes, including grandchild processes, even ones that were asynchronously spawned.

**Comment by ffulin** [comment]
@bburgin That's an interesting approach, but it seems that it would fail if any process suppresses the environment for its children. Any sandbox executable might do this for example. FASTBuild suppresses the environment of its child processes (paricularly important for distributed compilation), so it seems like Jenkins will fail to correctly terminate the process tree when running FASTBuild.

So far, it seems like the GetProcessTimes() check is the best fix - I can't think of any cases that doesn't handle correctly.

**Comment by bburgin** [comment]
Good points. But the env variable approach would allow FASTBuild to clean up orphaned child processes.
vctip.exe (experience data uploader), mspdbsrv.exe, others?

To make this work with the env manipulation that FASTBuild does:
The user could tell FASTBuild to pass along an FBUILD_CHILD_GROUP variable to child processes by importing it and then using it in the .Environment field of their Settings node:
Settings
{
    #if __WINDOWS__
    #import PATH
    #import TMP
    #import FASTBUILD_CHILD_GROUP
    .Environment = { 'PATH=$msvc_14_0_BasePath$/win64/x64/bin/amd64;$PATH$','TMP=$TMP$','SystemRoot=C:\Windows','FASTBUILD_CHILD_GROUP=$FASTBUILD_CHILD_GROUP$' }
    #endif
}

If the user passed thru the FASTBUILD_CHILD_GROUP, then FASTBuild would kill child processes since they would have the env variable and value that the parent FASTBuild is looking for.
Otherwise, FASTBuild would not kill the processes.

We could also make sandbox exes inherit this environment and clean those up, in the same way.

For Jenkins users, they could pass through the Jenkins build id env variable(s), if they wanted:
Settings
{
    #if __WINDOWS__
    #import PATH
    #import TMP
    #import FASTBUILD_CHILD_GROUP
    #import BUILD_ID
    #import JENKINS_NODE_COOKIE
    .Environment = { 'PATH=$msvc_14_0_BasePath$/win64/x64/bin/amd64;$PATH$','TMP=$TMP$','SystemRoot=C:\Windows','FASTBUILD_CHILD_GROUP=$FASTBUILD_CHILD_GROUP$','BUILD_ID=$BUILD_ID$','JENKINS_NODE_COOKIE=$JENKINS_NODE_COOKIE$' }
    #endif
}

Some additional questions:
Would the values of FASTBUILD_CHILD_GROUP, BUILD_ID, and JENKINS_NODE_COOKIE change too often and so invalidate caches too often?
The Jenkins BUILD_ID changes with every build, so that would invalidate caches.
One solution to this would be to allow the user to exclude certain env variables from the cache check.
So FASTBuild could remove those env variables from the env string before hashing the env.
This may also help with other use cases, if the user has other variables like this that change values often. Some examples I can think of:
1. the TMP folder variable may change across machines
2. the TMP folder variable may change when sandboxing
3. others?

If an exclude option does not make sense, then perhaps we could at least make the FASTBUILD_CHILD_GROUP value more deterministic across builds. Perhaps make it the path to the working directory of the FBuild.exe instance?

Thoughts?

**Comment by bburgin** [comment]
I am unclear what would happen if we killed mspdbsrv.exe while two fastbuilds were doing builds, so perhaps we would need a per exe exclusion list?

**Comment by ffulin** [comment]
Thanks for detailed thoughts. There is additional complexity to consider in all of this, which i'll brain dump below: :)

**Different Cancellation Scenarios**

There are two scenarios in which FASTBuild currently tries to cancel processes:
a) A single local task needs to be aborted, but others should remain untouched
b) You are aborting the build, and we want to stop all tasks

In case a, (for example: a remote builder returning a result and we abort the local race as per this bug, or a task failing or timeing out, but we're using --nostoponerror) we only want to abort the process tree for a single thread/job, and not any common sub-processes that might have been spawned as they can be doing work for the threads we don't want to terminate.

Even in case b), where we want to ultimately terminate all things that are running, having one thread cancel the shared sub-processes would cause spurious errors on the other threads which we would have to try and detect and suppress. Given the multi-threaded nature of task scheduling, there are likely to be race conditions and edge cases around. This is probably not insurmountable, but it'll be messy.

There are future cancellation scenarios I'd like to consider as well: At the moment, on a remote worker, if a client disconnects, the job will be completed and discarded. It would be better to cancel that task and immediately free up the resource for other jobs. As a remote worker can be compiling for multiple users are the same time, and these could be sharing mspdbsrv.exe instances, killing the process would cause unintended failures/.

**pdb corruption if terminating mspdbsrv.exe**

If we kill mspdbsrv.exe while it's in the middle of writing to a pdb, we risk corrupting pdb files, particularly when it's working for an unrelated task/thread to the one we are cancelling. Even if we know the mspdbsrv.exe is related to things we are cancelling, forcibly killing the process is likely to cause corruption. Unlike terminating a cl.exe (where we know the object is now invalid), the shared nature of the pdb files makes this problematic. 

**Is mspdbsrv.exe writing to the pdb synchronous?**

It's not clear whether cl.exe waits for mspdbsrv.exe to complete writing pdb info before exiting, or if it's somewhat asynchronous. There are no details I can find, but if pdbs can still be being written after a cl.exe exits, killing those mspdbsrv.exe processes could cause corrupt pdb files and/or discrepancies between FASTBuild and the on-disk state.

**Multiple instances of FASTBuild**

On a box where multiple concurrent FASTBuilds are running (in different directories for example, instances of mspdbsrv.exe can be shared between unrelated builds. One FASTBuild killing the mspdbsrv.exe could have unintended impact on other insatnces (errors/corruption).

**Undocumented _MSPDBSRV_ENDPOINT_**

There is an undocumented environment variable which can be used to isolate mspdbsrv.exe processes which might be useful in avoiding some of these problems (like multiple instance of FASTBuild running concurrently for example, or the FBuildWorker and FBuild sharing an instance).

Sadly, this doesn't help the case where a single FASTBuild must share a mspdbsrv.exe (because two cl.exe are updating the same pdb) but we only want to cancel one of them (the root scenario of this bug).

(Note that this might be worth doing completely outside of this bug/discussion to improve the behaviour of systems like Jenkins, or to cope with older/32-bit versions of mspdbsrv.exe that can run out of memory when linking in parallel.)

**Summary:**

Here's my current thinking:

 - If a process is shared and orphaned in the hierarchy, I think FASTBuild should assume this is deliberate or needed for things to operate safely and not touch that process. As bad as the design of mspdbsrv.exe is, it seems like we shouldn't fight it, or I think we risk causing more problems than we solve.
 - Ideally, I'd like to avoid FASTBuild having different cancellation modes, or having things you have to set in the config for it to work properly. At the very least, I'm leaning towards the process start time fix as the base behaviour, even if a more aggressive env-variable based thing might be useful as an option.

I'm going to keep pondering this problem. I appreciate any further thoughts!

**Comment by bburgin** [comment]
Thanks for explaining things. You've convinced me: I think you're right that a GetProcessTimes() fix is the best approach for the base behavior. I will leave it up to you and others on this thread to make the fix. Perhaps later we could add some env variable checking as a fallback to the base GetProcessTimes behavior.

**Comment by bburgin** [comment]
I went ahead and made the fix: https://github.com/fastbuild/fastbuild/pull/496

**Comment by ffulin** [comment]
I have submitted a fix for this issue using the GetProcessTimes() method discussed in this thread: https://github.com/fastbuild/fastbuild/commit/2fccad78ef5f7e5c59633c8d3b3e4acff275de90. This will be in the next release (v0.97)

I will leave this issue open until the next release.

@bburgin I'm sorry you spent your time on a fix in parallel to mine :( I ended up taking my fix because it solves some additional minor issues (leak of the hSnap snapshot handle), is a little simpler and also includes a basic process termination test.
